_docker/drupal/drupal-filesystem/web/modules/custom/rhd_common/rhd_common.sitemap.inc

```diff
@@ -39,9 +39,58 @@ function product_node_sitemap_entries(&$links) {
   // this add all the sub pages to the list of links and unsets the one with the node entry.
   $callback = function ($product_node_link, $product_node_link_index) use (&$links) {
     $product = Node::load($links[$product_node_link_index]['meta']['entity_info']['id']);
```

Comment:
@jordanpagewhite This is going to introduce a `N+1` selects problem e.g:

1. select the list of products
2. for each product, do at least another one select

This is a data access anti-pattern and seems to be impacting the performance of Cron pretty badly on your PR. More info on the n+1 select [here](https://bojanv91.github.io/posts/2018/06/select-n-1-problem).

I can see what you're trying to do from the code, so I'm wondering if there is a way we can query for each of the different product types e.g.:

1. select all product pages that are of the new type and details of their sub-pages in one query and add to sitemap.xml
   1. e.g "select from p from Product p where p.use_new_product_page=true`
   2. Is there a way we can include the logic to work out if it has the required subpages as well? 
2. select all product pages that are of the old type in one query and then add them to the `sitemap.xml`

Reply:
@jordanpagewhite Alternatively is is possible to a "pre-save" hook on the new product nodes that adds additional fields about the sub-pages of the node e.g. `hasDownloads`? This is a `boolean` field that is set to true when the product page is saved and the `field_downloads_page_content` collection is not empty. That means your query simply then has to load the product list and then look at the boolean fields for each entry in the list as to whether or not it should add additional entries to the `sitemap.xml`.